pkg/apply/applier.go

```diff
@@ -141,22 +144,66 @@ func (a *Applier) Run(ctx context.Context, invInfo inventory.InventoryInfo, obje
 			return
 		}
 
-		// Validate the resources to make sure we catch those problems early
-		// before anything has been updated in the cluster.
-		if err := (&object.Validator{
-			Mapper: mapper,
-		}).Validate(objects); err != nil {
-			handleError(eventChannel, err)
-			return
+		// Validate objects
+		var vErrors []error
+		validator := &validation.Validator{Mapper: mapper}
+		if err := validator.Validate(objects); err != nil {
+			vErrors = append(vErrors, err)
 		}
 
+		// Decide which objects to apply and which to prune
 		applyObjs, pruneObjs, err := a.prepareObjects(invInfo, objects, options)
 		if err != nil {
 			handleError(eventChannel, err)
 			return
 		}
 		klog.V(4).Infof("calculated %d apply objs; %d prune objs", len(applyObjs), len(pruneObjs))
 
+		// Sort applies into groups based on dependency ordering
```

Comment:
I pulled out some of the logic to a `validateAndSort` function.

> The original design was that more of of the logic for determining which resources should be applied/pruned and the order was in the Solver type.

Correct. This PR changes that, which makes the solver less of a solver. Another alternative would be to make the solver return a list of validation errors. Would that be preferable?

Reply:
Ok, I moved the sorting back into the Solver. It's slightly cleaner and fewer changes. To make it work, i had to inject InvalidIds into the Solver, because SortObjs needs the full input set, and then invalid objects need to be filtered afterwards, before the Task is created.

One interesting side effect of this is that the InvAddTask WILL include "invalid" objects that fail during sorting (depends-on, apply-time-mutation, and cycle errors), because the InvAddTask is built before sorting happens. This is actually exactly the same behavior as the current code. So it's not a regression. If we want to change this behavior, we can make the InvAddTask do filtering using the InvalidIds in the TaskContext.